packager/packager.cc

```diff
@@ -0,0 +1,731 @@
+     */
+    const PackageInfo *getPackageByFile(core::FileRef packageFile) const {
+        const auto &it = packageInfoByFile.find(packageFile);
+        if (it == packageInfoByFile.end()) {
+            return nullptr;
+        }
+        return it->second.get();
+    }
+
+    /**
+     * Given the mangled name for a package (e.g., Foo::Bar's mangled name is Foo_Bar_Package), return that package's
+     * info or nullptr if it does not exist.
+     */
+    const PackageInfo *getPackageByMangledName(core::NameRef name) const {
+        const auto &it = packageInfoByMangledName.find(name);
+        if (it == packageInfoByMangledName.end()) {
+            return nullptr;
+        }
+        return it->second.get();
+    }
+
+    /**
+     * Given a context, return the active package or nullptr if one does not exist.
+     */
+    const PackageInfo *getPackageForContext(core::Context ctx) const {
+        if (!finalized) {
+            Exception::raise("Cannot map files to packages until all packages are added and PackageDB is finalized");
+        }
+        // TODO(jvilk): Could use a prefix array to make this lookup more efficient.
+        auto path = ctx.file.data(ctx).path();
+        for (const auto &pkg : packages) {
+            if (absl::StartsWith(path, pkg->packagePathPrefix)) {
+                return pkg.get();
+            }
+        }
+        return nullptr;
+    }
+};
+
+void checkPackageName(core::Context ctx, ast::UnresolvedConstantLit &tree) {
+    auto constLit = &tree;
+    while (constLit != nullptr) {
+        if (absl::StrContains(constLit->cnst.data(ctx)->shortName(ctx), "_")) {
+            // By forbidding package names to have an underscore, we can trivially convert between mangled names and
+            // unmangled names by replacing `_` with `::`.
+            if (auto e = ctx.beginError(constLit->loc, core::errors::Packager::InvalidPackageName)) {
+                e.setHeader("Package names cannot contain an underscore");
+                auto replacement = absl::StrReplaceAll(constLit->cnst.data(ctx)->shortName(ctx), {{"_", ""}});
+                auto nameLoc = constLit->loc;
+                // cnst is the last characters in the constant literal
+                nameLoc.beginLoc = nameLoc.endLoc - constLit->cnst.data(ctx)->shortName(ctx).size();
+
+                e.addAutocorrect(core::AutocorrectSuggestion{
+                    fmt::format("Replace `{}` with `{}`", constLit->cnst.data(ctx)->shortName(ctx), replacement),
+                    {core::AutocorrectSuggestion::Edit{core::Loc(ctx.file, nameLoc), replacement}}});
+            }
+        }
+        constLit = ast::cast_tree<ast::UnresolvedConstantLit>(constLit->scope);
+    }
+}
+
+// Gets the package name in `tree` if applicable.
+PackageName getPackageName(core::MutableContext ctx, ast::UnresolvedConstantLit &constantLit) {
+    PackageName pName;
+    pName.loc = constantLit.loc;
+
+    {
+        auto current = &constantLit;
```

Comment:
The first thing that both `getPackageName` and `checkPackageName` do is to take the address of an alias to get a pointer. In both cases, I think it'd be cleaner to just pass the pointer in directly, and then not have to take the address (it happens to be fine in this case, but taking the address of things to get pointers is a really common footgun, so I'd like to avoid using it more widely).

Reply:
Fixed.